### ISSUE TITLE
Stronger trigger-fire typing

### ISSUE THREAD

**Issue by samskiter** [oxo42/stateless4j#34]
Hi, is it possible to have the trigger firing be more strongly typed/parameterised?

Specifically - to the params provided. It would be nice to force using of a certain number and type of args for a particular trigger. As I understand, currently I can 'fire' with or without any params. I then need to deal with that in my onEntryBlocks

I had a think about how you might achieve this:

1. After making a `TriggerWithParameters` - you call a method on it to generate a concrete instance of the trigger. I.e. you have a trigger factory of sorts that you use when firing events.

2. Use a reflection method similar to Retrofit. Supply an interface of events that is parsed down into `TriggerWithParameters` objects. This might be a Java 8 only option because to configure the onEntryBlock you might have to use a method refs to identify the event methods.

**Comment by samskiter** [comment]
Update: I've just found the `setTriggerParameters` method which allows what I proposed in option 1. This might be a good case for some more example code/docs.

Perhaps making the constructor of `TriggerWithParameters` package-private might help - I was instantiating one...

**Comment by Ktar5** [comment]
I agree, this should be in the documentation. I might write something up on this later today.
